### PR TITLE
COTR-105 Implemented shop functionality

### PR DESCRIPTION
**Pull request by disusen**
Implemented basic shop functionality which allows the user to spend his coins on various weapons found in the shop.

### CODE REVIEW

**Review by Elijuszek — CHANGES_REQUESTED** [review]
- Error when switching weapons
- Files which haven't been touched shouldn't be committed (`.tscn` files)
- When work is of large scope and working across multiple classes changes should be staged into multiple commits (for example two commits for: changes in `Shop.cs` and changes in `WeaponsManager.cs`)

File: Scenes/Actors/Player/WeaponsManager.cs [hunk 1]
```diff
@@ -60,16 +56,26 @@ public void Initialize(Team.Teams setTeam, Weapon weapon)
     /// Add weapon to currently carrying weapon list
     /// </summary>
     /// <param name="index">Index to put in to</param>
-    public void AddWeapon(Weapon weapon, int index)
+    public bool AddWeapon(Weapon weapon, int index)
```
Comment (Karolinskis) [inline Scenes/Actors/Player/WeaponsManager.cs:59]: Due to changing return from `void` to `bool` the summary above this method is incorrect add `<returns>`

File: Scenes/Actors/Player/WeaponsManager.cs [hunk 2]
```diff
@@ -84,18 +90,13 @@ public Weapon GetCurrentWeapon()
     /// <summary>
     /// Change the current weapon the user is holding
     /// </summary>
-    public void ChangeWeapon()
+    public void ChangeWeapon(int indexToChangeTo)
```
Comment (Karolinskis) [inline Scenes/Actors/Player/WeaponsManager.cs:93]: Incorrect summary add <param name="indexToChangeTo">

File: Scenes/UI/HUD/GUI.cs [hunk 3]
```diff
@@ -194,5 +196,36 @@ private void ChangeItem(int index, Weapon weapon = null)
...
+        SwitchItem(1);
+    }
+
+    /// <summary>
+    /// Switches weapon to the third one in weaponsManager.
+    /// </summary>
+    private void ItemPressed3()
+    {
+        SwitchItem(2);
+    }
```
Comment (Karolinskis) [inline Scenes/UI/HUD/GUI.cs:230]: This code is unmaintainable. This should probably be one method called `ItemPressed` that you pass and argument in which you tell which button was pressed.

**Review by Karolinskis — CHANGES_REQUESTED** [review]

**Review by Elijuszek — CHANGES_REQUESTED** [review]

**Review by Karolinskis — CHANGES_REQUESTED** [review]
If nothing was bought from the store, the player should spawn with no weapon. And the inventory should be empty.

**Review by Karolinskis — CHANGES_REQUESTED** [review]
Bug mentioned in https://github.com/Nekompiliuoja/Champions-of-the-Realm/pull/91/commits/f417eedfae778c73d4378ad736635058b80565ac still persists

**Review by Karolinskis — CHANGES_REQUESTED** [review]

**Review by Karolinskis — CHANGES_REQUESTED** [review]
When starting a new game in the debuger an error pops up

**Comment by Karolinskis** [comment]
> When starting a new game in the debuger an error pops up

---
E 0:00:08:0416   Godot.NativeInterop.NativeFuncs.generated.cs:1535 @ void Godot.NativeInterop.NativeFuncs.godotsharp_pusherror(Godot.NativeInterop.godot_string& ): Calling Idle method from weapon class
  <C++ Source>   modules/mono/glue/runtime_interop.cpp:1303 @ godotsharp_pusherror()
  <Stack Trace>  Godot.NativeInterop.NativeFuncs.generated.cs:1535 @ void Godot.NativeInterop.NativeFuncs.godotsharp_pusherror(Godot.NativeInterop.godot_string& )
                 GD.cs:351 @ void Godot.GD.PushError(System.String )
                 Weapon.cs:42 @ void ChampionsOfTheRealm.Weapon.Idle()
                 Melee.cs:22 @ void ChampionsOfTheRealm.Melee._Ready()
                 Node.cs:1801 @ Boolean Godot.Node.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CanvasItem.cs:1371 @ Boolean Godot.CanvasItem.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 Node2D.cs:516 @ Boolean Godot.Node2D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 ChampionsOfTheRealm.Weapon_ScriptMethods.generated.cs:61 @ Boolean ChampionsOfTheRealm.Weapon.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 ChampionsOfTheRealm.Melee_ScriptMethods.generated.cs:75 @ Boolean ChampionsOfTheRealm.Melee.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_variant_call_error* , Godot.NativeInterop.godot_variant* )
                 Godot.NativeInterop.NativeFuncs.generated.cs:332 @ void Godot.NativeInterop.NativeFuncs.godotsharp_method_bind_ptrcall(IntPtr , IntPtr , System.Void** , System.Void* )
                 NativeCalls.cs:5859 @ void Godot.NativeCalls.godot_icall_3_655(IntPtr , IntPtr , IntPtr , Godot.NativeInterop.godot_bool , Int32 )
                 Node.cs:600 @ void Godot.Node.AddChild(Godot.Node , Boolean , Godot.Node+InternalMode )
                 Map.cs:127 @ void ChampionsOfTheRealm.Map.SpawnPlayer()
                 Map.cs:53 @ void ChampionsOfTheRealm.Map._Ready()
                 Node.cs:1801 @ Boolean Godot.Node.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CanvasItem.cs:1371 @ Boolean Godot.CanvasItem.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 Node2D.cs:516 @ Boolean Godot.Node2D.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 ChampionsOfTheRealm.Map_ScriptMethods.generated.cs:96 @ Boolean ChampionsOfTheRealm.Map.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_variant_call_error* , Godot.NativeInterop.godot_variant* )
                 Godot.NativeInterop.NativeFuncs.generated.cs:332 @ void Godot.NativeInterop.NativeFuncs.godotsharp_method_bind_ptrcall(IntPtr , IntPtr , System.Void** , System.Void* )
                 NativeCalls.cs:5859 @ void Godot.NativeCalls.godot_icall_3_655(IntPtr , IntPtr , IntPtr , Godot.NativeInterop.godot_bool , Int32 )
                 Node.cs:600 @ void Godot.Node.AddChild(Godot.Node , Boolean , Godot.Node+InternalMode )
                 LoadingScreen.cs:96 @ void ChampionsOfTheRealm.LoadingScreen.AddScene()
                 LoadingScreen.cs:71 @ void ChampionsOfTheRealm.LoadingScreen.InitializeResourceLoader()
                 ChampionsOfTheRealm.LoadingScreen_ScriptMethods.generated.cs:54 @ Boolean ChampionsOfTheRealm.LoadingScreen.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name& , Godot.NativeInterop.NativeVariantPtrArgs , Godot.NativeInterop.godot_variant& )
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(IntPtr , Godot.NativeInterop.godot_string_name* , Godot.NativeInterop.godot_variant** , Int32 , Godot.NativeInterop.godot_variant_call_error* , Godot.NativeInterop.godot_variant* )

**Review by Karolinskis — APPROVED** [review]

File: Scenes/UI/HUD/GUI.cs [hunk 4]
```diff
@@ -52,6 +53,13 @@ public override void _Ready()
         itemArray[0] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems");
         itemArray[1] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems2");
         itemArray[2] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems3");
+
+        // Assigns each button to SwitchItem method.
+        for (int i = 0; i < itemArray.Length; i++)
+        {
+            int currentIndex = i;
+            itemArray[i].Pressed += () => SwitchItem(currentIndex);
```
Comment (Karolinskis) [inline Scenes/UI/HUD/GUI.cs:61]: Refactor this

**Review by Karolinskis — CHANGES_REQUESTED** [review]

File: Scenes/UI/HUD/GUI.cs [hunk 5]
```diff
@@ -55,10 +55,10 @@ public override void _Ready()
         itemArray[2] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems3");
 
         // Assigns each button to SwitchItem method.
-        for (int i = 0; i < itemArray.Length; i++)
+        foreach (Button button in itemArray)
         {
-            int currentIndex = i;
-            itemArray[i].Pressed += () => SwitchItem(currentIndex);
+            int ButtonIndex = Array.IndexOf(itemArray, button);
+            button.Pressed += () => SwitchItem(ButtonIndex);
```
Comment (Karolinskis) [inline Scenes/UI/HUD/GUI.cs:61]: This anonymous lambda expression effects code readability in a bad way, please Refactor this part.

**Review by Karolinskis — CHANGES_REQUESTED** [review]

File: Scenes/UI/HUD/GUI.cs [hunk 6]
```diff
@@ -52,6 +53,13 @@ public override void _Ready()
         itemArray[0] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems");
         itemArray[1] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems2");
         itemArray[2] = GetNode<Button>("HUD/MarginContainer/Rows/BottomRow/InventoryContainer/MarginContainer/VBoxContainer/IntentoryItems3");
+
+        // Assigns each button to SwitchItem method.
+        foreach (Button button in itemArray)
+        {
+            int ButtonIndex = Array.IndexOf(itemArray, button);
+            button.Pressed += () => SwitchItem(ButtonIndex);
+        }
```
Comment (Karolinskis) [inline Scenes/UI/HUD/GUI.cs:62]: Refactor this whole part, don't assign methods through code you can do this through the editor.
image

**Review by Karolinskis — CHANGES_REQUESTED** [review]

File: Scenes/UI/HUD/GUI.cs [hunk 7]
```diff
@@ -5,6 +5,7 @@ namespace ChampionsOfTheRealm;
 /// </summary>
 public partial class GUI : Control
 {
+    [Signal] public delegate void WeaponSwitchedEventHandler(int index);
```
Comment (Karolinskis) [inline Scenes/UI/HUD/GUI.cs:8]: Why did you add this delegate? It's not needed

**Review by Karolinskis — CHANGES_REQUESTED** [review]

File: Scenes/Actors/Player/WeaponsManager.cs [hunk 8]
```diff
@@ -59,17 +55,28 @@ public void Initialize(Team.Teams setTeam, Weapon weapon)
...
+    /// <returns>True if weapon added succesfully, false otherwise</returns>
+    public bool AddWeapon(Weapon weapon, int index)
     {
+        if (this.GetChildCount() > 2)
+        {
+            return false;
+        }
+        if (index >= weapons.Length)
+        {
+            Array.Resize(ref weapons, index + 1);
```
Comment (Karolinskis) [inline Scenes/Actors/Player/WeaponsManager.cs:69]: What's the point of resizing the array when the array is constantly size of 3?

**Review by Karolinskis — CHANGES_REQUESTED** [review]

File: Scenes/UI/HUD/GUI.cs [hunk 9]
```diff
@@ -186,13 +188,21 @@ private void ChangeItem(int index, Weapon weapon = null)
         {
             itemArray[index].Icon = null;
         }
-
+        itemArray[index].Disabled = false;
+        itemArray[index].FocusMode = Godot.Control.FocusModeEnum.All;
```
Comment (Elijuszek) [inline Scenes/UI/HUD/GUI.cs:192]: Remove this unnecessary line. By default buttons `focusMode = All`.

**Review by Elijuszek — CHANGES_REQUESTED** [review]

File: Scenes/UI/HUD/GUI.cs (hunk 9 above)
Comment (Elijuszek) [reply Scenes/UI/HUD/GUI.cs:192]: Okay makes sense

File: Scenes/Actors/Player/WeaponsManager.cs (hunk 8 above)
Comment (disusen) [reply Scenes/Actors/Player/WeaponsManager.cs:69]: The array has a maximum of size 3 for items attached through the editor, if any items are added dynamically, like in the shop then array will have to be resized.

**Review by Karolinskis — APPROVED** [review]

**Review by Elijuszek — APPROVED** [review]
LFTM!

**Pull request merged**